numpyro/distributions/flows.py

```diff
@@ -79,3 +79,46 @@ def log_abs_det_jacobian(self, x, y, intermediates=None):
         else:
             log_scale = intermediates
             return log_scale.sum(-1)
+
+
+class BlockNeuralAutoregressiveTransform(Transform):
```

Comment:
In Pyro, this class is named `BlockAutoregressive` and to construct a transform, we use
```
# create the transform directly
transform = BlockAutoregressive(input_dim, hidden_factors=[8, 8], residual=None)
# I think this is the way to register parameters of the transform...
pyro.module("bnarn", transform)
```

In NumPyro, to construct a transform, we use
```
# create a nn module, which is a pair of init_fn, apply_fn
bnarn = BlockNeuralAutoregressiveNN(input_dim, hidden_factors=[8, 8], residual=None)
# then we register the above module
bnarn = numpyro.module('bnarn', bnarn, (input_dim,))
# finally, create the transform
transform = BlockNeuralAutoregressiveTransform(bnarn)
```
Beside the names are different, this design not only follows AutoregressiveNN and InverseAutoregressionTransform, but also be specific to NumPyro/JAX. In particular, we declare a nn module separately from the transform because in NumPyro, transform is a class while a nn module is a pair of `(init_fn, apply_fn)`.

I think most users will use the autoguide AutoBNAFNormal (not available yet) instead of the above nn/transform. In that case, the above difference will not be exposed to users. Both NumPyro/Pyro should have the same interface
```
guide = AutoBNAFNormal(model, num_flows=2, hidden_factors=[8, 8])
```

cc @martinjankowiak

Reply:
I think we have just been using the transforms, in which case I wonder if it makes sense to keep the NN interface hidden and have the transforms directly construct the required NNs internally in `__init__`. That will avoid this indirection of first creating NNs, calling module, and then passing these to construct the transform. This will be a deviation from the Pyro design so its probably worth a separate discussion.

Reply:
Yup, it will be cleaner to combine those two steps but the signature of a transform (with call, inverse, logdet) is different from a stax nn layer. I guess we can create TransformModule as in Pyro and let `module` handler know that this is a special nn layer. Something like
```
# create a TransformModule, this is not a Transform because no param is available yet!
# signature of TransformModule is: `call(params, x)`, `inverse(params, y)`, `logdet(params, x, y)`
transform = BlockNeuralAutoregressiveTransform(*bnarn)
# register TransformModule to `module` handler, after this line, we get a real Transform!
transform = numpyro.module('transform', transform, (input_dim,))
```

Let me make a separate issue for this so we can discuss it further. I don't like the pattern in the above sketch. :(

Reply:
Thanks for creating the issue. This isn't a blocker for the current PR, but we should discuss the design trade-offs separately.